<a href="https://colab.research.google.com/github/dondreojordan/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module1/Jordan_Dondre_Group_TL_Maggie_DS17_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

## 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

#Use the automobile dataset

In [0]:
from scipy import stats
import pandas as pd
import numpy as np

In [9]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
#Loading in the '85 Automobile Dataset

--2020-06-08 21:02:55--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.1s    

2020-06-08 21:02:55 (194 KB/s) - ‘imports-85.data’ saved [25936/25936]



#Fix the column headers

In [137]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ration', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
# Create column headers from the imports-85.names file.

df = pd.read_csv('imports-85.data', names=column_headers)
# Read in the dataframe.

print(df.shape)
df.head()
# Print head and ensure column_headers was added correctly.

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

#Make sure NaNs are used to indicate missing values

In [90]:
df = df.replace({'?':np.NaN}) # Let's clean up the data

# Replace the values
df.isnull().sum()

symboling              0
normalized-losses     41
make                   0
fuel-type              0
aspiration             0
num-of-doors           2
body-style             0
drive-wheels           0
engine-location        0
wheel-base             0
length                 0
width                  0
height                 0
curb-weight            0
engine-type            0
num-of-cylinders       0
engine-size            0
fuel-system            0
bore                   4
stroke                 4
compression-ration     0
horsepower             2
peak-rpm               2
city-mpg               0
highway-mpg            0
price                  4
dtype: int64

In [28]:
#Let's see how many NaN values are in which column.
df.isnull().sum()

symboling              0
normalized-losses     41
make                   0
fuel-type              0
aspiration             0
num-of-doors           2
body-style             0
drive-wheels           0
engine-location        0
wheel-base             0
length                 0
width                  0
height                 0
curb-weight            0
engine-type            0
num-of-cylinders       0
engine-size            0
fuel-system            0
bore                   4
stroke                 4
compression-ration     0
horsepower             2
peak-rpm               2
city-mpg               0
highway-mpg            0
price                  4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. **Note** that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [115]:
Rand_state_30 = df.sample(30, random_state=30)

print()
Rand_state_30.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

In [70]:
df['highway-mpg'].mean() # Mean of highway-mpg.

30.75121951219512

In [112]:
Rand_state_30['highway-mpg'].mean() # Mean of highway-mpg.

32.46666666666667

>> Can you say that they are different?

By definition, the sample is different from popular value of highway-mpg, however, more efficient and  similar in retrospect.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of **10 cars**, test his claim and report your results.

In [127]:
test10 = Rand_state_30.sample(10)
print(test10.shape)
test10.head()

(10, 26)


symboling normalized-losses           make  ... city-mpg highway-mpg  price
72           3               142  mercedes-benz  ...       16          18  35056
176         -1                65         toyota  ...       27          32  10898
20           0                81      chevrolet  ...       38          43   6575
178          3               197         toyota  ...       20          24  16558
21           1               118          dodge  ...       37          41   5572

[5 rows x 26 columns]

In [128]:
test10['highway-mpg'].describe()

count    10.000000
mean     31.900000
std       8.723531
min      18.000000
25%      25.000000
50%      32.000000
75%      40.000000
max      43.000000
Name: highway-mpg, dtype: float64

Imagine that you now have the capacity to test **100 cars** on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [139]:
test100 = df.sample(100, random_state=30)
test205= df.sample(205, random_state=30) #Let's say he sold every car on the lot.

print(test100.shape)
print(test205.shape)

(100, 26)
(205, 26)


In [131]:
test10['highway-mpg'].mean()

31.9

In [130]:
test100['highway-mpg'].mean()

30.64

In [140]:
test205['highway-mpg'].mean()

30.75121951219512

>>Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Both datasets lead to different conclusions about the salesman's claim of seel cars averaging to 35 miles per gallon on the highway.  When testing the claim with 10 cars you will notice the data is limited in comparison with 100 cars or all 205 cars on the lot (say he sold every car on the lot). The more the population the more accurate the result.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
from scipy import stats
import pandas as pd
import numpy as np

In [145]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-09 00:32:56--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-06-09 00:32:57 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [148]:
column_headers = ['Class Name', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers) #read in .csv file

df.head() #Verify column_headers

Class Name  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [149]:
df.isnull().sum()

Class Name                                0
handicapped-infants                       0
water-project-cost-sharing                0
adoption-of-the-budget-resolution         0
physician-fee-freeze                      0
el-salvador-aid                           0
religious-groups-in-schools               0
anti-satellite-test-ban                   0
aid-to-nicaraguan-contras                 0
mx-missile                                0
immigration                               0
synfuels-corporation-cutback              0
education-spending                        0
superfund-right-to-sue                    0
crime                                     0
duty-free-exports                         0
export-administration-act-south-africa    0
dtype: int64

In [150]:
df = df.replace({'y': 1, 'n': 0, '?':np.NaN}) # Let's clean up the data

# Replace the values
df.head()

Class Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [159]:
df.isnull().sum()

Class Name                                  0
handicapped-infants                        12
water-project-cost-sharing                 48
adoption-of-the-budget-resolution          11
physician-fee-freeze                       11
el-salvador-aid                            15
religious-groups-in-schools                11
anti-satellite-test-ban                    14
aid-to-nicaraguan-contras                  15
mx-missile                                 22
immigration                                 7
synfuels-corporation-cutback               21
education-spending                         31
superfund-right-to-sue                     25
crime                                      17
duty-free-exports                          28
export-administration-act-south-africa    104
dtype: int64

In [151]:
df.columns

Index(['Class Name', 'handicapped-infants', 'water-project-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-groups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

In [0]:
dem = df[df['Class Name'] == 'democrat']
rep = df[df['Class Name'] == 'republican']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

#Using hypothesis testing, find an issue that **democrats** support more than **republican**s with p < 0.01

#2) 2 sample example
>>Null Hypothesis: the two sample means are equal. Democrats support this bill at the **same** rate that Republicans do.

>>Alternative Hypothesis: They support this bill at different rates

>>Confidence Level: 95%

In [153]:
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [154]:
print(rep['handicapped-infants'].mean())
print(dem['handicapped-infants'].mean())

0.18787878787878787
0.6046511627906976


#Conclusion:

>>Based on a t-statistic of 16.34 and a p-value of 0 (we know becuase of 9.95e-47), we **reject** the null hypothesis that replublicans and democrats support this bill at the same rate, and suggest the alternative, that republicans and democrats show different levels of support for this bill.

Do Democrats or Republicans support this bill more?

**Democrats**

#Using hypothesis testing, find an issue that **republicans** support more than democrats with p < 0.01

In [155]:
stats.ttest_ind(rep['superfund-right-to-sue'], dem['superfund-right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [156]:
print(rep['superfund-right-to-sue'].mean())
print(dem['superfund-right-to-sue'].mean())

0.8607594936708861
0.2896825396825397


#Conclusion:

>>Based on a t-statistic of 13.51 and a p-value of 0 (we know becuase of 1.227e-34), we **reject** the null hypothesis that replublicans and democrats support this bill at the same rate, and suggest the alternative, that republicans and democrats show different levels of support for this bill.

Do Democrats or Republicans support this bill more?

**Republicans**

#Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)


In [157]:
stats.ttest_ind(rep['water-project-cost-sharing'], dem['water-project-cost-sharing'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [158]:
print(rep['water-project-cost-sharing'].mean())
print(dem['water-project-cost-sharing'].mean())

# Notice that out of all those 387 Representatives that voted
# it was almost a split decision. 

0.5067567567567568
0.502092050209205


#Conclusion:

>>Based on the p-value of .92, we do not reject the the null hypothesis that republicans and democrats support this bill at the same rate, and **do not** suggest the alternative hypothesis stating that republicans and democrats show different levels of support for this bill. 

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:
cols = df.columns[1:]

for col in cols[1:]:



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE